lib/CodeGen/MemoryAllocator.cpp

```diff
@@ -138,49 +154,626 @@ uint64_t MemoryAllocator::allocate(uint64_t size, Handle handle,
+      currSegAddrStop = currSegAddrStart + buffInfoArray[currSegId].size;
+
+      // Verify if this placement overlaps with all the other segments.
+      // Note that this verification with all the previous segments is required
+      // because the previous segments can overlap between themselves.
+      bool overlap = false;
+      for (size_t ovrSegIdx = 0; ovrSegIdx < prevSegAddr.size(); ovrSegIdx++) {
+        // Check overlap.
+        overlap = overlap || intervalsOverlap(currSegAddrStart, currSegAddrStop,
+                                              prevSegAddr[ovrSegIdx].first,
+                                              prevSegAddr[ovrSegIdx].second);
+        // Early break if overlaps.
+        if (overlap) {
+          break;
+        }
+      }
+
+      // If no overlap than we found the solution for the placement.
+      if (!overlap) {
+        break;
+      }
+    }
+
+    // Update maximum used size.
+    usedSizeMax = std::max(usedSizeMax, currSegAddrStop);
+
+    // If max available memory is surpassed with the new segment then we stop
+    // the allocation and return early.
+    if (usedSizeMax > memorySize) {
+      return MemoryAllocator::npos;
+    }
+
+    // Allocate current segment.
+    Segment currSeg(currSegAddrStart, currSegAddrStop);
+    idSegMap.insert(std::make_pair(currSegId, currSeg));
+
+    // Update buffer liveness information.
+    for (size_t allocIdx = buffInfoArray[currSegId].timeStart;
+         allocIdx < buffInfoArray[currSegId].timeStop; allocIdx++) {
+      // Update total live size.
+      liveInfoArray[allocIdx].size -= buffInfoArray[currSegId].size;
+      // Update total live IDs.
+      auto &allocIds = liveInfoArray[allocIdx].idList;
+      auto it = std::find(allocIds.begin(), allocIds.end(), currSegId);
+      assert(it != allocIds.end() && "Buffer ID not found for removal!");
+      allocIds.erase(it);
+    }
+  }
+
+  // Verify again that all the buffers were allocated.
+  for (size_t allocIdx = 0; allocIdx < liveInfoArray.size(); allocIdx++) {
+    assert(liveInfoArray[allocIdx].size == 0 &&
+           "Not all buffers were allocated!");
+    assert(liveInfoArray[allocIdx].idList.empty() &&
+           "Not all buffers were allocated!");
+  }
+
+  return usedSizeMax;
+}
+
+void MemoryAllocator::mapHandlesToIds(
+    const std::list<Allocation> &allocList,
+    std::unordered_map<Handle, size_t> &handleToIdMap,
+    std::vector<Handle> &idToHandleMap) {
+  size_t buffNum = allocList.size() / 2;
+  handleToIdMap.clear();
+  idToHandleMap = std::vector<Handle>(buffNum);
+  size_t id = 0;
+  for (const auto &alloc : allocList) {
+    // We only map the Handles of ALLOCs.
+    if (alloc.alloc_) {
+      handleToIdMap[alloc.handle_] = id;
+      idToHandleMap[id] = alloc.handle_;
+      id++;
+    }
+  }
+  assert(id == buffNum && "Inconsistent Handle to ID mapping!");
+}
+
+uint64_t MemoryAllocator::allocateAll(const std::list<Allocation> &allocList,
```

Comment:
Nit: This function is still pretty big. Don't know if it is possible to split it into smaller parts.

Reply:
Yeah I`m not sure I can split it further.